### PR TITLE
[MCKIN-9301] BB-728 - Modify LMS APIs to enable OAuth access to problem response reports

### PR DESCRIPTION
**Pull request by giovannicimolin**
This PR enables OAuth access of the endpoints necessary to generate, get status and download problem response reports.

**JIRA tickets**: [OSPR-2983](https://openedx.atlassian.net/browse/OSPR-2983), [BB-728](https://tasks.opencraft.com/browse/BB-728)

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: 25/01/2019 (end of Sprint 186).

**Testing instructions**:

1. Run tests related to this new endpoint:
```
paver test_system -t lms/djangoapps/instructor/tests/test_api.py
```
2. [Go to the `DemoX edX Demonstration Course` as **staff** user.
](http://lms.mcka.local/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-data_download)
3. Check if the generation of problem response reports is working correctly via the instructor interface.
4. [Create an Oauth client here if one does not exists for testing.](http://localhost:18000/admin/oauth2/client)
5. Use the Oauth2 endpoint to get an auth token (cUrl, Postman or other tool).
6. Check get_response_report endpoint. Example:
Request:
```
POST /courses/course-v1:edX+DemoX+Demo_Course/instructor/api/get_problem_responses
Body:
{
    "problem_location": "block-v1:edX+DemoX+Demo_Course+type@problem+block@c554538a57664fac80783b99d9d6da7c"
}
```
Response:
```
{
  "status": "The problem responses report is being created. To view the status of the report, see Pending Tasks below."
}
```
7. Check list_instructor_tasks endpoint. Example:
Request:
```
POST /courses/course-v1:edX+DemoX+Demo_Course/instructor/api/list_instructor_tasks
Body: {}
```
Response:
```
{
  "tasks": [
    {
      "status": "Incomplete", 
      "task_type": "grade_course", 
      "task_id": "4f78e78c-3d66-4265-af21-ca75c43036a3", 
      "created": "2019-01-15T18:00:14.088198+00:00", 
      "task_input": "{}", 
      "duration_sec": "unknown", 
      "task_message": "No status information available", 
      "requester": "staff", 
      "task_state": "PROGRESS"
    }
  ]
}
```
8. Check list_report_downloads endpoint. Example:
Request:
```
POST /courses/course-v1:edX+DemoX+Demo_Course/instructor/api/list_report_downloads
Body: {}
```
Response:
```
{
  "downloads": [
    {
      "url": "/media/bb6d638a296059742509a0d319ddf8456b6dbf9a/edX_DemoX_Demo_Course_student_state_from_block-v1_edX%2BDemoX%2BDemo_Course%2Btype%40problem%2Bblock%40c554538a57664fac80783b99d9d6da7c_2019-01-18-1353.csv", 
      "link": "<a href=\"/media/bb6d638a296059742509a0d319ddf8456b6dbf9a/edX_DemoX_Demo_Course_student_state_from_block-v1_edX%2BDemoX%2BDemo_Course%2Btype%40problem%2Bblock%40c554538a57664fac80783b99d9d6da7c_2019-01-18-1353.csv\">edX_DemoX_Demo_Course_student_state_from_block-v1_edX+DemoX+Demo_Course+type@problem+block@c554538a57664fac80783b99d9d6da7c_2019-01-18-1353.csv</a>", 
      "name": "edX_DemoX_Demo_Course_student_state_from_block-v1_edX+DemoX+Demo_Course+type@problem+block@c554538a57664fac80783b99d9d6da7c_2019-01-18-1353.csv"
    }
  ]
}
```

**Author notes and concerns**:

1. I'm not sure if it's better to move these endpoints to `/api/v1/instructor`. I've left them in the same place as before for now.

**Reviewers**
- [ ] @xitij2000 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @giovannicimolin! I've created [OSPR-2983](https://openedx.atlassian.net/browse/OSPR-2983) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

**Comment by natabene** [comment]
@giovannicimolin Thank you for your contribution, please let me know once it is ready to be looked at.

**Comment by giovannicimolin** [comment]
@natabene This still needs to go through internal review. I'll ping you once it's ready for review by edX.

File: lms/djangoapps/instructor/views/api.py [hunk 1]
```diff
@@ -220,6 +220,15 @@ def wrapped(*args, **kwargs):  # pylint: disable=missing-docstring
     return decorator
 
 
+class IsCourseStaff(permissions.BasePermission):
```
Comment (xitij2000) [inline lms/djangoapps/instructor/views/api.py]: Could you move this to the same file that has the IsCourseStaffInstructor permission, and add a corresponding test?

(openedx/core/lib/api/permissions.py)

File: lms/djangoapps/instructor/views/api.py [hunk 2]
```diff
@@ -2297,50 +2320,77 @@ def list_email_content(request, course_id):  # pylint: disable=unused-argument
...
-    List instructor tasks.
+    **Use Cases**
+
+        Lists currently running instructor tasks
 
-    Takes optional query paremeters.
+    **Parameters**
         - With no arguments, lists running tasks.
-        - `problem_location_str` lists task history for problem
+        - `problem_location_str` lists task history for problemself.
```
Comment (xitij2000) [inline lms/djangoapps/instructor/views/api.py]: problemself -> problem

File: lms/djangoapps/instructor/views/api.py [hunk 3]
```diff
@@ -2297,50 +2320,77 @@ def list_email_content(request, course_id):  # pylint: disable=unused-argument
...
-    if student and not problem_location_str:
-        return HttpResponseBadRequest(
-            "unique_student_identifier must accompany problem_location_str"
-        )
+    @cache_control(no_cache=True, no_store=True, must_revalidate=True)
+    def post(self, request, course_id):
+        """
+        List instructor tasks.
+        """
+        course_id = SlashSeparatedCourseKey.from_deprecated_string(course_id)
```
Comment (xitij2000) [inline lms/djangoapps/instructor/views/api.py]: Could you change instances of SlashSeparatedCourseKey... to CourseKey...
See discussion [here](https://github.com/edx/edx-platform/pull/18975#discussion_r222773392)

**Pull request merged**